### PR TITLE
[varLib] Far faster featureVars

### PR DESCRIPTION
**Pull request by behdad**
Motivated by https://github.com/fonttools/fonttools/issues/1370

### CODE REVIEW

**Comment by behdad** [comment]
While the merging of duplicate regions fixed https://github.com/fonttools/fonttools/issues/1370, I like to rewrite the algorithm to be roughly O(num-out-regions x num-input-regions) region operations instead of current more than O(2^num-input-regions) region operations.

File: Lib/fontTools/varLib/featureVars.py [hunk 1]
```diff
@@ -29,22 +29,62 @@ def addFeatureVariations(font, conditionalSubstitutions):
     The minimum and maximum values are expressed in normalized coordinates.
 
     A Substitution is a dict mapping source glyph names to substitute glyph names.
+
+    Example:
+
+    # >>> f = TTFont(srcPath)
+    #>>> condSubst = [
```
Comment (justvanrossum) [inline Lib/fontTools/varLib/featureVars.py]: Inconsistent formatting with the previous line.

**Comment by behdad** [comment]
> This is a sensible refactoring. Curious how the algorithm will improve.

See the test.  Current algorithm generates 2**n-1 regions whereas only 2*n-1 is needed.

**Comment by behdad** [comment]
Okay!! @justvanrossum this is ready for review.  Someone needs to inspect failing test for correctness of new output and update expectation.

**Comment by behdad** [comment]
From commit logs of the two main commits:
```
[varLib.featureVars] Rewrite algorithm

Whereas previous algorithm had exponential running time and return
value size, new one has quadratic.

For featureVars_test.py test case, for example, which is a pathological
test case of n sliding intervals, the number of output intervals of
various algorithms are:

- Previous algorithm: 2**n - 1
- New algorithm: n*(n-1)/2
- Optimal algorithm: 2*n - 1

Ie, we go from exponential to quadratic, whereas in this case the optimal
solution is linear.

Running time of said test, for n=20, goes from over 20s, to 0.06s.

The algorithm can be improved.  The overlayBox() function currently does
not try to shrink the remainder box.  Doing that will probably bring us
to optimal solution for this test case.

Fixes #1372

One test is failing.  Needs to be investigated that new output is correct,
and test expectations updated.
```
and later:
```
[varLib.featureVars] Improve new algorithm to its full potential

See 0283b1f

By finishing overlayBox() to cull remainder box when possible, we generate far fewer
spurious boxes.

For the featureVars_test.py test case, this now generates optimal output, that is, 2*n - 1
boxes.

Running time for n=100 is up from 0.06s to 0.16s, but output size goes down from
5050 boxes to the optimal 199.

Fixes #1372 again
```

**Comment by anthrotype** [comment]
Bravo! 👏

**Comment by behdad** [comment]
Meh.  Found bugs.  Reworking.  The optimal solution might not happen.

**Comment by behdad** [comment]
Bugs resolved.  Same results.

**Comment by justvanrossum** [comment]
That one test failure seems harmless, in that two FeatureVariationRecords appear to have swapped places. As far as I can see these are non overlapping regions, so that change in order should not make a difference in behavior.

Would be nice to have explicit tests for complex configurations of regions.

**Pull request merged**